circuit-runtime/src/commonMain/kotlin/com/slack/circuit/runtime/Navigator.kt

```diff
@@ -3,14 +3,21 @@
 package com.slack.circuit.runtime
 
 import androidx.compose.runtime.Stable
+import com.slack.circuit.runtime.screen.PopResult
 import com.slack.circuit.runtime.screen.Screen
 
-/** A basic navigation interface for navigating between [screens][Screen]. */
+/** A Navigator that only supports [goTo]. */
 @Stable
-public interface Navigator {
+public interface GoToNavigator {
   public fun goTo(screen: Screen)
+}
+
+/** A basic navigation interface for navigating between [screens][Screen]. */
+@Stable
+public interface Navigator : GoToNavigator {
+  public override fun goTo(screen: Screen)
 
-  public fun pop(): Screen?
+  public fun pop(result: PopResult? = null): Screen?
```

Comment:
Can you unpack the first line a bit for me more?

> Combing popping and indicating a return value restricts how both can/must be used. 

For other options, I considered a few! Including some parts of that prototype.
1. Something similar to `ActivityResultContract`, where both the target and the result are generic types.
    - Benefits are that they're reusable, so common contracts can be easily defined and reused.
    - Cons are that they're limited to specific 1:1 pairings and the target screen doesn't have to abide by the contract.
2. Something like `AnsweringScreen<T : PopResult>`, which would be a slimmer version of #1. Target screen still doesn't have to abide by it, superficial value. Cons are the target screen also can only return that one result, rather than multiple.
3. Silently alter the previous screen with a result when popping back, such that the recreated presenter is created with a new screen with a populated result. I couldn't really think of a way to fit this into an API though, every screen would need to expose some sort of copy() function that could receive a result. This also prevents screens from supporting multiple different result types.

For returning the result, nowhere else seemed really appropriate than the `Navigator`. I know your prototype did this with a new `ScreenResultHandler`, but it doesn't feel right. It requires doing two actions that always go together (set the result then pop()) and adds a whole new top-level type to the Presenter.Factory. It also requires moving result handling to the `(Navigable)CircuitContent` entry point, which is too removed from the actual caller to do anything within the presenter that called it. It also was set up with a lot of extra infrastructure for returning results out of circuit, which we ended up solving separately internally with (IMO) simpler APIs.

Some of the prototype stuff has similar echoes in here! ScreenResult/PopResult interface, hosting it in the record, etc. Across the different approaches, one common denominator was that there was never really a way to keep type safety on the other end, or at least not without a lot of tedium and hard-to-lint developer stories to do it. This feels like the right middle ground of API ergonomics and functionality.